### DISCUSSION TITLE
next function in Strategy not taking steps

### DISCUSSION THREAD

**Discussion by jjphung** [kernc/backtesting.py#266] [Q&A] [answered]
Hello all,

For some reason my Strategy class isn't iterating through the data using the `next` function.

```
    # ... 

    def init(self):
        self.c = self.data.Close.s
        h = self.data.High.s
        l = self.data.Low.s
        self.tenkansens = self.I(self.tenkansen, h, l).s
        self.kijunsens = self.I(self.kijunsen, h, l).s
        self.spanAs = self.I(self.spanA, self.tenkansens, self.kijunsens).s
        self.spanBs = self.I(self.spanB, h, l).s
        print(self.tenkansens)

    def next(self):
        print(self.tenkansens.iloc[-1])

```
The print statement in `next` just only prints out the same number over and over again, while the print statement in `init` does indeed show unique values.

I also checked all 4 of these series to see if there were at least 2 elements in there that weren't NaN, so I should see at least see one iteration in the `next` function.

Am I missing anything here? Let me know if you need anymore details.

**Comment by kernc** [comment, marked as answer]
You are assigning:
```py
self.tenkansens = self.I(self.tenkansen, h, l).s     # type: pd.Series
```
that is, the indicator but transformed into `pd.Series`, which makes it invisible to Backtest for iterative lengthening.

If you instead do it like this, it should work:
```py
def init(self):
    self.tenkansens = self.I(self.tenkansen, h, l)   # type: _Indicator

def next(self):
    print(self.tenkansens.s.iloc[-1])
```

**Reply by jjphung** [reply]
Awesome! Did not know it would be invisible. Thanks.

**Reply by kernc** [reply]
Yeah, we use this to recognize the "indicators" saved on `self`:
https://github.com/kernc/backtesting.py/blob/73e1534428fc2db9e681030862133b5067508520/backtesting/backtesting.py#L1137-L1140
If you have a better idea, welcome!

**Reply by jjphung** [reply]
Sounds good :)

I do have a follow up question though: 

```
    def init(self):
        self.c = self.data.Close

    def next(self):
        if self.tenkansens.s.iloc[-1] < self.c:
            # Sample signal triggered
```

For the code sample above, would this be the correct way of checking if the current close price is above the tenkansen's price? Or is there a better way? Thanks again!

**Reply by kernc** [reply]
I think it would be. Also this:
```py
    def next(self):
        if self.data.Close > self.tankansens:
#        if self.data.Close[-1] > self.tankansens[-1]:  # Same, but relatively more performant
            ...
```
i.e. without saving `self.c` separately because `self.data` is always available, and without costly conversion to `pd.Series` just to throw it away.
